### PR TITLE
Major Architecture Changes

### PR DESCRIPTION
**Pull request by imcodingideas**
# Major Architecture Changes

I started to move the server side to separate logic from routes, and actions. I'm still working on fixing https://github.com/freeCodeCamp/classroom-mode/issues/51 https://github.com/freeCodeCamp/classroom-mode/issues/44 but the core is there. 

Fixes # https://github.com/freeCodeCamp/classroom-mode/issues/51 https://github.com/freeCodeCamp/classroom-mode/issues/44

## Type of change

- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

# Testing

There are still several tests that need to be updated. 

Frontend: 
- [ ]  src/__test__/AddStudentForm.test.js

Backend: 
- [x]  Get Tests to Run

Scraper via: scraper.js:
- [X] should return an error if scraper has a non-200 status code
- [X] should return no errors if scraper has a 200 status code
- [ ] should compute correct number of inactive days
- [X] should compute correct number of inactive days when user has no history
- [X] should compute correct number of inactive days when user was active today
- [ ] should respect last updated_at when calculating last daysInactive

POST /add_student via: addStudentTest.js:
- [x] should return an error if student name is absent
- [x] should return an error if email in invalid
- [x] should receive an error message if scraper returns error
- [ ]  should call the mongoose save method when all fields are valid
- [X]  should not call the mongoose save method when username is invalid
- [ ]  should save completedChallengesCount and completedChallengesTitle when saving students
- [x] should find a Student by username. **new** 
- [X] should returns new submissions count and titles when student completedChallengesCount is undefined - **Deprecated**. 

GET /students via: getStudentTest.js:
- [X]  should return 200
- [ ]  should fetch data from mongo DB
- [ ]  should return a 200 and an empty array if the database is empty
- [X]  should look up student github username
- [ ]  should returns new submissions count and titles

Validating student records: studentModelValidationTest.js: **New**.
- [X]  requires a name
- [X]  requires a students's name longer than 2 characters
- [X]  disallows invalid records from being saved

### CODE REVIEW

**Comment by imcodingideas** [comment]
I'm still working on this, but Im going to take a break for a little bit.

**Comment by imcodingideas** [comment]
I'm having an issue figuring out why `Days Inactive`, and `New Submissions` are not showing. I see the data there right before we send the `students object` in the `/server/controllers/studentController.js`

I know: `fetchUserInfoFromFCC` works.

**Comment by user512** [comment]
@imcodingideas, thank you for your contribution, your contribution provided a better code structure, great!

For your issue on not getting `Day Inactive` and `New Submissions`, My guess is the student schema prevent you from modifying the mongoose object.

I tried adding `dayInactive` and `newSubmissionsCount` into `Student.js` and I was able to display both values. However, it sounds weird that we have to modify the student schema since we are not saving those value to db, I wonder if there is another way to solve this problem.

**Comment by imcodingideas** [comment]
@user512 I think we can something that's called virtual values. I will add that today!

**Comment by user512** [comment]
You are right, @imcodingideas, we might not want to use virtual in this case.
I am thinking we can use `toObject()` to convert it into a plain JavaScript object so we can modify it and render to the client.
http://mongoosejs.com/docs/api.html#document_Document-toObject

**Comment by imcodingideas** [comment]
@user512 thanks for the PR. I will work on this tomorrow. It's been a crazy Thursday/Friday.

**Comment by imcodingideas** [comment]
@user512 I could use some help here on the tests. I've disabled the ones that are broken but the functionality does work. Just the tests are wrong.

There were tests that were duplicated, and I added a few more.

**Comment by imcodingideas** [comment]
@user512 Can you merge this, and I will continue to work on fixing these one at a time with new features i.e. the edit. Creating a higher order function for the loader when you save.

**Comment by imcodingideas** [comment]
I will move validation of an email out of the controller now that it's in the model, than work on the delete.

File: server/test/scraper.js [hunk 1]
```diff
@@ -11,188 +11,192 @@ afterEach(function() {
...
+    profileImage: 'https://avatars3.githubusercontent.com/u/6780322?v=4',
+    location: 'San Francisco',
+    completedChallenges: []
+  });
 
-      scraper.fetchUserInfoFromFCC('testUser', function(err, results) {
-        expect(err).to.equal(true);
-        expect(results.error).to.be.a('object');
-        done();
+  it('should return an error if scraper has a non-200 status code', function(done) {
```
Comment (honoriovega) [inline server/test/scraper.js]: Possible duplicate test

**Pull request merged**